### PR TITLE
TDD Schedule Query Algorithm

### PR DESCRIPTION
**Pull request by dterr**
MVP for Query Algorithm through test-driven development.
Currently not commented nor styled, which will occur in next commits.
Meant as a check-in and proof of work.

### CODE REVIEW

**Comment by dterr** [comment]
MVP complete.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java [hunk 1]
```diff
@@ -28,6 +29,16 @@
...
+
+  /**
+  * A comparator for sorting events by their start time in ascending order.
+  */
+  public static final Comparator<Event> ORDER_BY_START_TIME = new Comparator<Event>() {
+    @Override
+    public int compare(Event a, Event b) {
+      return Long.compare(a.when.start(), b.when.start());
+    }
+  };
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java:40]: new line below

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java [hunk 2]
```diff
@@ -28,6 +29,16 @@
   private final TimeRange when;
   private final Set<String> attendees = new HashSet<>();
 
+
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java]: Remove extra line.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -15,9 +15,179 @@
...
+*      Calculate those free windows
+*      Calculate intersection
+*/
 public final class FindMeetingQuery {
+
+  /**
+  * Function query
+  * This method searches for common free windows in the schedules of the attendees needed. 
+  * Optional attendees for the meeting are also considered and accommodated if possible.
+  * @param request The object containing the requesting event and details
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:46]: add new line before @param

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -15,9 +15,179 @@
...
+  * @param events A list of all of the events and attendees for the day
+  * @return A list of possible time frames that satisfy the request.
+  *
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: remove extra new line

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,179 @@
...
+    ArrayList<String> optionals = new ArrayList<>(request.getOptionalAttendees());
+    long duration = request.getDuration();
+    
+    if (duration > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    } else if (events.isEmpty() || guests.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    if (!optionals.isEmpty()) {
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit:  you can have the empty case first. That's easier to read.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -270,5 +273,127 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
+
+  @Test
+  public void completelyBusyOptionalAttendee() {
+    /* Test three attendees who have different schedules,
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Can you also illustrate the windows similar to previous tests so it's easier to understand and also more consistent?

Comment (mwo-2020) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Also, previous tests use "//" style comments. Maybe also fix to be consistent.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -15,9 +15,179 @@
...
+  */
+  private Collection<TimeRange> meetingWithOptionals
+      (Collection<String> mandatory, Collection<String> optionals, 
+      long duration, Collection<Event> events) {
+        
+    ArrayList<TimeRange> mandatoryWindows = (ArrayList<TimeRange>) communalFreeWindows(mandatory, duration, events);
+    ArrayList<TimeRange> optionalWindows = (ArrayList<TimeRange>) communalFreeWindows(optionals, duration, events);
+
+    if (mandatoryWindows.isEmpty()) {
+      return optionalWindows;
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:89]: When there are no available window for the mandatory guests, should this return empty list?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java (hunk 1 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java:40]: Addressed: d864a97

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java (hunk 2 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java]: Addressed: d864a97

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:46]: Addressed: 1d79835

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Addressed: 00853d2

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Addressed: 1d79835

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 6 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Addressed: 45ce2a9

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Addressed: 1d79835

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:89]: Addressed: 00853d2

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java [hunk 8]
```diff
@@ -28,6 +29,16 @@
   private final TimeRange when;
   private final Set<String> attendees = new HashSet<>();
 
+  /**
+  * A comparator for sorting events by their start time in ascending order.
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java]: Align the *  (similar to line 24).  Here and in the rest of the files.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -15,9 +15,181 @@
...
+  * @param sortedUnavailableTimes A list of busy times ordered for the day.
+  * @param duration The desired length of the meeting.
+  * @return A list of communal free times that are sufficiently long.
+  */
+  private Collection<TimeRange> findAppropiateFreeWindows(Collection<TimeRange> sortedUnavailableTimes, long duration) {
+    List<TimeRange> freeWindows = new ArrayList<>();
+
+    int previousEventEnd = 0;
+    for (TimeRange curWhen : sortedUnavailableTimes) {
+      if (curWhen.start() <= previousEventEnd) { //Overlapping events
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Add space after //

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 10]
```diff
@@ -270,5 +273,150 @@ public void notEnoughRoom() {
 
     Assert.assertEquals(expected, actual);
   }
+
+  @Test
+  public void completelyBusyOptionalAttendee() {
+    // Test three attendees who have different schedules, 
+    // including one of whom is optional with an all-day event. 
+    //  Events  : |--A--|
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: remove extra space before Events

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java [hunk 11]
```diff
@@ -92,6 +92,31 @@ public boolean overlaps(TimeRange other) {
...
+  /**
+  * Returns the overlapping time window between two TimeRange elements.
+  * If there is no overlap, throws exception.
+  */
+  public TimeRange getOverlap(TimeRange other) {
+    if (!this.overlaps(other)) {
+      throw new IllegalArgumentException("TimeRange arguments do not overlap.");
+    }
+
+    //Should I use .start() or .start here?
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java]: Both are okay. Maybe use start() to be consistent.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 12]
```diff
@@ -270,5 +273,150 @@ public void notEnoughRoom() {
...
+
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TIME_1115AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void onlyUnavailableOptionalAttendees() {
```
Comment (mwo-2020) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:394]: Should this scenario return no options?

Comment (mwo-2020) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:394]: Sorry I was wrong. Re-read the walkthrough. This is fine.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java (hunk 8 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/Event.java]: Addressed: 80d9acb

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Addressed: 344c43b.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 10 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Addressed: c3ec025

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java (hunk 11 above)
Comment (dterr) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/TimeRange.java]: Good point, thank you.
233ab83

**Review by mwo-2020 — APPROVED** [review]
Looks good! Congrats on finishing the starter project!

**Pull request merged**